### ISSUE TITLE
Translate Cassandra's CQL unit tests into Scylla's cql-pytest framework

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#7722]
The Apache Cassandra project has in their main source repository a directory `test/unit/org/apache/cassandra/cql3` [(see in gitbox)](https://gitbox.apache.org/repos/asf?p=cassandra.git;a=tree;f=test/unit/org/apache/cassandra/cql3;h=ed14ff67eb12bce93c9ece4e7febda73fede5247;hb=refs/heads/trunk) a nice collection of tests - over 50,000 lines of code - for the various CQL features, entities, operations, and so on.

Let's translate these tests into our new **cql-pytest** framework.

The translation work should be fairly straightforward - both frameworks rely on the ability to run CQL commands, and check their results, and not on calls to internals as most "traditional" unit tests.

Note that these tests will be a bit slower than our usual cql-pytest and alternator tests (though still much faster than dtest). Most of our own tests were written to heavily use fixtures to avoid creating and recreating the same table in every test - but in Cassandra's cql3 tests every test function creates and deletes its own new temporary table. This slows down the tests because table creation and deletion is by far the slowest operation in most those tests. Issue #7620 means it further slows down when we give each temporary table a random name; We can work around that problem by keeping a cache of _deleted table names_ and reusing such a name instead of inventing a new random name in every test.  In the long run, many tests that use the same table schema should be able to use the same table without creating and deleting it many times (by using fixtures). However, this will require careful changes to the test code, because many test functions currently assume the table is empty at the beginning of every test function, so we'll probably not want to do these changes on the first step.

Some of these tests have comments that they have been migrated from dtests. When we translate such a test, let's check if it exists in our own dtest repository, and if it does, remove it from there - we don't need it in both and dtest is much slower. 

CCing people who have expressed interest in this in the past: @eliransin @dekimir @penberg @avikivity

**Comment by nyh** [comment]
Note that I'm slowly translating all of Cassandra's CQL-related tests to `test/cql-pytest/cassandra_tests`.
I'm not sure if I should close this issue now, or wait until I finished translating the last test (which can take a while...).

**Comment by psarna** [comment]
GitHub has this fancy option of checkboxes, so you could leverage that in the issue description to keep track of the progress:
- [x] filtering tests
- [ ] collection tests
... etc

**Comment by psarna** [comment]
ref: https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/about-task-lists

**Comment by nyh** [comment]
> GitHub has this fancy option of checkboxes, so you could leverage that in the issue description to keep track of the progress:
> 
>     * [x]  filtering tests
> 
>     * [ ]  collection tests
>       ... etc

Thanks, but I don't think I want to do this. I can make a long list of Cassandra source files and mark them here each time I translate one, but it's too easy to forget updating this list, and just "ls -R test/cql-pytest/cassandra-tests" can get you this list.

**Comment by dekimir** [comment]
>  I'm not sure if I should close this issue now, or wait until I finished translating the last test (which can take a while...).

This is a general problem with huge work items that aren't broken down into manageable scopes.  Another is that it's hard to prioritize among individual steps (especially when there are multiple gigantic-scope bugs).  Maybe those of us urging you to file multiple narrow bugs are on to something here... ;)

**Comment by dekimir** [comment]
That said, I still think translating Cassandra tests (and our own unit tests for CQL features) to cql-pytest is the right path forward.

**Comment by nyh** [comment]
@dekimir I don't know what you are suggesting. If there are at least 40 untranslated test files, containing hundreds of tests. I've been translating them slowly, and sending a patch each time I finish translating a file (plus opening many small issues for the individual bugs I discover). I don't understand what you'd like me to do differently.

In the meantime I'm leaving this issue open just as a reminder that we have **not** finished this translation work, despite the fact we already have a directory `test/cql-pytest/cassandra-tests`. The existence of this directory just means I started this translation - not that I finished it.

**Comment by dekimir** [comment]
> @dekimir I don't know what you are suggesting.

I'm suggesting filing narrower bugs that can be knocked out in a week or so.  Larger chunks of work should be broken down.  I'm suggesting filing only those bugs that we know we'll work on in the next month or two; this will make the list size manageable, force us to prioritize, and keep us flexible when priorities inevitably change with time.

> In the meantime I'm leaving this issue open just as a reminder that we have not finished this translation work,

That's fine, but I think we should refrain from filing gigantic-scope bugs going forward.  Breaking work down is a valuable discipline to acquire.

**Comment by nyh** [comment]
> I'm suggesting filing narrower bugs that can be knocked out in a week or so.

Yes, but this is already what I always do - except for this issue. What do you recommend I do for **this** issue?

Close it? I can.

What I can't do is "break it down". This is not practical nless there are other volunteers who want to help me translate tests, and then we really need some sort of coordination so we don't translate the same tests concurrently.

**Comment by dekimir** [comment]
> > I'm suggesting filing narrower bugs that can be knocked out in a week or so.
> 
> Yes, but this is already what I always do - except for this issue.

And the `assert` bug.  And the UNSET_VALUE bug.  And the reversed-types bug.  See a pattern?

> What do you recommend I do for **this** issue?
> 
> Close it? I can.

No, I already said it's fine to keep it open now that it exists.

> What I can't do is "break it down".

I don't see why not.  There is any number of ways this work can be split.

> This is not practical nless there are other volunteers who want to help me translate tests, and then we really need some sort of coordination so we don't translate the same tests concurrently.

I volunteer to help, and I propose the traditional method of coordination: file multiple bugs and assign them to different people. :bulb:

**Comment by nyh** [comment]
> > > I'm suggesting filing narrower bugs that can be knocked out in a week or so.
> > 
> > 
> > Yes, but this is already what I always do - except for this issue.
> 
> And the `assert` bug. And the UNSET_VALUE bug. And the reversed-types bug. See a pattern?

No, I don't see "a pattern", sorry. Each of the bugs you mentioned here took you less than one day to fix. How is this relevant your request that a bug take less than one **week** to fix? Yes, you can give me other reasons why you wanted to split those bugs, and I gave you reasons why you shouldn't have split them. But limiting the scope to "bugs that can be knocked out in a week or so" is simply not relevant here.
 
> > This is not practical nless there are other volunteers who want to help me translate tests, and then we really need some sort of coordination so we don't translate the same tests concurrently.
> 
> I volunteer to help, and I propose the traditional method of coordination: file multiple bugs and assign them to different people. bulb

I see. I didn't know you were volunteering.
If we're two people, let's split the directories. I'll continue working in `test/cql-pytest/cassandra_tests/validation/entities` (there are still tons of stuff to be done there), and you pick a different directory. If you think it makes sense you can open a new bug for the directory you start translating.

**Comment by dekimir** [comment]
> > And the `assert` bug. And the UNSET_VALUE bug. And the reversed-types bug. See a pattern?
> 
> No, I don't see "a pattern", sorry. Each of the bugs you mentioned here took you less than one day to fix.

Except I didn't fix the whole bug, just a limited scope.  For instance, we're closing #7740 (CQL prepared statements incomplete support for "unset" values) though we know there are likely more instances of such incomplete support.  It's the exact same problem you're seeing in this issue.  (As a pedantic note: #7740 and #7868 (as interpreted by you on the mailing list) are of unknown scope, so not necessarily gigantic.  But it's equivalent for the purposes of discussing breaking work down into limited-scope chunks.)

And you yourself wrote [here](https://github.com/scylladb/scylla/issues/7871#issuecomment-754776605) that there is the same discussion going on in multiple bugs.

> If we're two people, let's split the directories. I'll continue working in `test/cql-pytest/cassandra_tests/validation/entities` (there are still tons of stuff to be done there), and you pick a different directory. If you think it makes sense you can open a new bug for the directory you start translating.

OK.

**Comment by mykaul** [comment]
Thanks @nyh for the summary. As we've discussed, I'm not sure what the best path forward is - continue in the 'known' areas or explore more 'unknowns' - fix existing issues or translate more to uncover more issues (or get reassurance that features work). 
I wonder if looking at some of the subdirectories content and translating few will give us some value
